logstash-core/spec/logstash/compiler/compiler_spec.rb

```diff
@@ -560,18 +560,92 @@ def compose(*statements)
             end
 
             describe "'=='" do
-              let(:expression) { "[foo] == 5"}
+              let(:lhs) { fail(NotImplementedError) }
+              let(:expression) { "[foo] == #{rhs}"}
 
-              it "should compile correctly" do
-                expect(c_expression).to ir_eql(j.eEq(j.eEventValue("[foo]"), j.eValue(5.to_java)))
+              context 'when RHS represents an integer' do
+                let(:rhs) { '5' }
+                it "should compile correctly" do
+                  expect(c_expression).to ir_eql(j.eEq(j.eEventValue("[foo]"), j.eValue(5.to_java)))
+                end
+              end
+
+              context 'when RHS represents an empty array' do
+                let(:rhs) { '[]' }
+                it "should compile correctly" do
+                  expect(c_expression).to ir_eql(j.eEq(j.eEventValue("[foo]"), j.eValue([])))
+                end
+              end
+
+              # multi-item arrays are not supported because a single-item arrays in this context are
+              # syntactically ambiguous with _selector_s, which take a higher precedence; if/when support
+              # is added for unambiguously declaring a single-item array as a condition clause, this spec
+              # can be removed.
+              context 'when RHS represents a multi-item array' do
+                let(:rhs) { '["foo","bar"]'}
+                it "should fail to compile with a helpful message" do
+                  expect { c_expression }.to raise_error do |error|
+                    expect(error).to be_a_kind_of(RuntimeError)
+                    expect(error.message).to include("Invalid Config: Comparison exprs do not support non-empty arrays")
+                  end
+                end
+              end
+
+              # equality with a regexp is not something people _intend_ to do, and
+              # is an example of a valid `rvalue` that _parses_ but doesn't have semantic meaning.
+              context 'when RHS represents a Regexp' do
+                let(:rhs) { '/foo/' }
+                it "should fail to compile with a helpful message" do
+                  expect { c_expression }.to raise_error do |error|
+                    expect(error).to be_a_kind_of(RuntimeError)
+                    expect(error.message).to include("Invalid Config: Unsupported RegExp value on right-hand side of boolean expr")
```

Comment:
Maybe this should be more actionable for the user -- 

Suggestion:

> `RegExp value <value> may only be used with the =~ or !~ operator, for example: #{lval} =~ #{rval}`

thoughts?

Reply:
This spec is an example of what the "uncaught" `rvalue` would produce, and adding specific call-outs to known-not-related nodes is a bit of a slippery slope that I was trying to avoid (e.g., If we were to pass a valid `method_call`, which has no underlying support, a similar message would be generated, except it would have `MethodCall` instead of `RegExp`).

I agree that we could probably have better messaging for `RegExp`, even though it never should belong as the RHS of an equality or comparison operation; I still feel that this patch is a net improvement, but am willing to add in the extra warning for `RegExp`, since it is more likely to get hit:

With this patch:
~~~
Invalid Config: Unsupported RegExp value on right-hand side of boolean expr; line 24 column 18 (byte 228) in `[foo] == /bar/`
~~~

Previously:
~~~
Failed to parse right-hand side of conditional [foo] == /bar/
~~~